### PR TITLE
Fish tanks, Episode 5 : Fish tanks strike back [Ready]

### PR DESCRIPTION
**Pull request by ShiftyRail**
"They did not know it was impossible, so they did it" - Mark Twain

## Scope of this PR
Resurrects #16428, fixing of the comments, eventual conflicts,  and ZAS problem.

Tested, but still needs a few things to do

## Gallery

https://streamable.com/y9try 
https://streamable.com/i2qiw

## Bugs

- [x] ~Why won't the fishtank break properly ?~
- [x] ~Why has the sprite always been the same during my tests ?~
- [x] ~Why do fishtanks sometimes let gases pass through ?~
- [x] ~Why can't I disassemble the fishtank properly ?~
- [x]  ~And probably some other shit I forgot to clean up~
- [x] Why doesn't examine work when there fish in the tank ?

Thanks to jakmak for doing a lot of work on this, and to Blithering for spritework.

:cl:
- rscadd: Added fishtanks! Construct your own fishtank with glasses, metal and a circuit board. Breed fish with eggs from cargo. Catch fish, make your own fish-based recipes! Sprites courtesy of Blithering. Original fishtanks on Paradise courtesy of TheFalseIncarnate, original work on fishtanks courtesy of jakmak.

### CODE REVIEW

File: code/modules/fish/fish_items.dm [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	desc = "A tiny net to capture fish with. It's a death sentence!"
+	icon = 'icons/obj/fish_items.dmi'
+	icon_state = "net"
+	slot_flags = SLOT_BELT
+	throwforce = 0
+	w_class = 2
+	throw_speed = 3
+	throw_range = 7
+
+	/obj/item/weapon/fishtools/fish_net/suicide_act(mob/user)			//"A tiny net is a death sentence: it's a net and it's tiny!" https://www.youtube.com/watch?v=FCI9Y4VGCVw
```
Comment (MadmanMartian) [inline code/modules/fish/fish_items.dm]: Why's this tabbed?

File: code/modules/fish/fish_items.dm [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	icon = 'icons/obj/fish_items.dmi'
+	icon_state = "brush"
+	slot_flags = SLOT_BELT
+	throwforce = 0
+	w_class = 2
+	throw_speed = 3
+	throw_range = 7
+	attack_verb = list("scrubbed", "brushed", "scraped")
+
+	/obj/item/weapon/fishtools/fish_tank_brush/suicide_act(mob/user)
```
Comment (MadmanMartian) [inline code/modules/fish/fish_items.dm]: Why's this tabbed?

File: code/modules/fish/fish_items.dm [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+/obj/item/weapon/reagent_containers/food/snacks/shrimp
+	name = "shrimp"
+	desc = "A single raw shrimp."
+	icon = 'icons/obj/seafood.dmi'
+	icon_state = "shrimp_raw"
+	filling_color = "#FF1C1C"
+
+/obj/item/weapon/reagent_containers/food/snacks/shrimp/New()
+		..()
```
Comment (MadmanMartian) [inline code/modules/fish/fish_items.dm]: Why the doubletab?

File: code/modules/fish/fish_items.dm [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+		src.bitesize = 1
+
+/obj/item/weapon/reagent_containers/food/snacks/feederfish
+	name = "feeder fish"
+	desc = "A tiny feeder fish. Sure doesn't look very filling..."
+	icon = 'icons/obj/seafood.dmi'
+	icon_state = "feederfish"
+	filling_color = "#FF1C1C"
+
+	/obj/item/weapon/reagent_containers/food/snacks/feederfish/New()
```
Comment (MadmanMartian) [inline code/modules/fish/fish_items.dm]: Why is this tabbed?

File: code/modules/fish/fish_items.dm [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+	icon_state = "shark"
+	hitsound = 'sound/weapons/bite.ogg'
+	force = 3
+
+/obj/item/weapon/fish/shark/attackby(var/obj/item/O, var/mob/user as mob)
+	if(istype(O, /obj/item/weapon/wirecutters))
+		to_chat(user, "You rip out the teeth of \the [src.name]!")
+		new /obj/item/weapon/fish/toothless_shark(get_turf(src))
+		new /obj/item/stack/teeth/shark(get_turf(src))
+		qdel(src)
```
Comment (MadmanMartian) [inline code/modules/fish/fish_items.dm]: You'll want to make sure either the shark's on a surface, or is dropped by the user if they're holding it

File: code/modules/fish/fish_items.dm [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+	name = "catfish"
+	desc = "Apparently, catfish don't purr like you might have expected them to. Such a confusing name!"
+	icon_state = "catfish"
+
+/obj/item/weapon/fish/catfish/attackby(var/obj/item/O, var/mob/user as mob)
+	if(O.sharpness_flags & SHARP_BLADE)
+		to_chat(user, "You carefully clean and gut \the [src].")
+		new /obj/item/weapon/reagent_containers/food/snacks/catfishmeat(get_turf(src))
+		new /obj/item/weapon/reagent_containers/food/snacks/catfishmeat(get_turf(src))
+		qdel(src)
```
Comment (MadmanMartian) [inline code/modules/fish/fish_items.dm]: Same as the shark attackby, make sure it's either on a surface, or is dropped if the user is holding it

File: code/modules/fish/fish_items.dm [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+/obj/item/stack/teeth/shark/New()
+	src.pixel_x = rand(-5,5)
+	src.pixel_y = rand(-5,5)
+
+/obj/item/weapon/fish/catfish
+	name = "catfish"
+	desc = "Apparently, catfish don't purr like you might have expected them to. Such a confusing name!"
+	icon_state = "catfish"
+
+/obj/item/weapon/fish/catfish/attackby(var/obj/item/O, var/mob/user as mob)
```
Comment (MadmanMartian) [inline code/modules/fish/fish_items.dm]: No need for the 'as mob' in this or other instances

File: code/modules/fish/fish_tank.dm [hunk 8]
```diff
@@ -0,0 +1,687 @@
...
+
+//////////////////////////////
+//		VERBS & PROCS		//
+//////////////////////////////
+
+/obj/machinery/fishtank/verb/toggle_lid_verb()
+	set name = "Toggle Tank Lid"
+	set category = "Object"
+	set src in view(1)
+	if (usr.incapacitated() || !usr.Adjacent(src))
```
Comment (MadmanMartian) [inline code/modules/fish/fish_tank.dm:122]: May want to check for ghosts, too, so you don't get spooky lid toggling

File: code/modules/fish/fish_tank.dm [hunk 9]
```diff
@@ -0,0 +1,687 @@
...
+
+/obj/machinery/fishtank/proc/toggle_lid(var/mob/living/user)
+	lid_switch = !lid_switch
+	update_icon()
+
+/obj/machinery/fishtank/verb/toggle_light_verb()
+	set name = "Toggle Tank Light"
+	set category = "Object"
+	set src in view(1)
+	if (usr.incapacitated() || !usr.Adjacent(src))
```
Comment (MadmanMartian) [inline code/modules/fish/fish_tank.dm:134]: check for ghosts, or don't if you want spooky goings on

File: code/modules/fish/fish_items.dm (hunk 5 above)
Comment (ShiftyRail) [reply code/modules/fish/fish_items.dm]: ~I haven't tested this code yet (hence DNM) but I'm relatively certain this part isn't working as intended~
Never mind it works

File: code/modules/fish/fish_tank.dm [hunk 10]
```diff
@@ -0,0 +1,687 @@
...
+		var/dead_fish = null
+		dead_fish = fish_items_list[caught_fish]	//Locate the appropriate fish_item for the caught fish
+		if(!dead_fish)								//No fish_item found, possibly due to typo or not being listed. Do nothing.
+			return
+		kill_fish(caught_fish)						//Kill the caught fish from the tank
+		user.visible_message("[user.name] harvests \a [caught_fish] from \the [src].", "You scoop \a [caught_fish] out of \the [src].")
+		new dead_fish(get_turf(user))				//Spawn the appropriate fish_item at the user's feet.
+
+/obj/machinery/fishtank/proc/destroy(var/deconstruct = FALSE)
+	to_chat(world, "being destroyed")
```
Comment (MadmanMartian) [inline code/modules/fish/fish_tank.dm]: DEBUG TEXT, PLS REMOVE

File: code/modules/fish/fish_tank.dm [hunk 11]
```diff
@@ -0,0 +1,687 @@
...
+	if(!deconstruct)
+		to_chat(world, "Brutally")												//Check if we are deconstructing or breaking the tank
```
Comment (MadmanMartian) [inline code/modules/fish/fish_tank.dm]: DEBUG TEXT, PLS REMOVE

File: code/modules/fish/fish_tank.dm [hunk 12]
```diff
@@ -0,0 +1,687 @@
...
+		for(var/i in  shard_count)														//Produce the appropriate number of glass shards
+			var/obj/item/weapon/shard/S = new /obj/item/weapon/shard
+			S.forceMove(get_turf(src))
+		if(water_level)
+			to_chat(world, "Spilling_water")											//Spill any water that was left in the tank when it broke
```
Comment (MadmanMartian) [inline code/modules/fish/fish_tank.dm]: DEBUG TEXT, PLS REMOVE

**Review by MadmanMartian — CHANGES_REQUESTED** [review]

File: code/modules/fish/fish_tank.dm (hunk 10 above)
Comment (ShiftyRail) [reply code/modules/fish/fish_tank.dm]: Beat you to it

File: code/modules/fish/fish_tank.dm [hunk 13]
```diff
@@ -0,0 +1,685 @@
...
+/obj/machinery/fishtank/examine(mob/user)
+	..()
+	var/examine_message = list()
+	//Approximate water level
+
+	examine_message += "Water level: "
+
+	if(water_level == 0)
+		examine_message += "\The [src] is empty! "
+	else if(water_level < water_capacity * 0.1)
```
Comment (MadmanMartian) [inline code/modules/fish/fish_tank.dm]: May want to test this, I've a feeling if it's full it will print out every damn one of these when examining.

File: code/modules/fish/fish_tank.dm [hunk 14]
```diff
@@ -0,0 +1,685 @@
...
+		user.visible_message("<span class='danger'>\The [user] smashes into \the [src]!</span>")
+		playsound(loc, 'sound/effects/Glasshit.ogg', 100, 1)
+		check_health()
+
+/obj/machinery/fishtank/attackby(var/obj/item/O, var/mob/user as mob)
+	//Welders repair damaged tanks on help intent, damage on all others
+	if(iswelder(O))
+		var/obj/item/weapon/weldingtool/W = O
+		if(user.a_intent == I_HELP)
+			if(W.isOn())
```
Comment (MadmanMartian) [inline code/modules/fish/fish_tank.dm]: Don't we use silicate sprayers for repairing glass?

Also you'll want to use the use proc or equivalent so it consumes some of the welder fuel/silicate when repairing

File: code/modules/fish/fish_tank.dm [hunk 15]
```diff
@@ -0,0 +1,685 @@
...
+		kill_fish(caught_fish)						//Kill the caught fish from the tank
+		user.visible_message("[user.name] harvests \a [caught_fish] from \the [src].", "You scoop \a [caught_fish] out of \the [src].")
+		new dead_fish(get_turf(user))				//Spawn the appropriate fish_item at the user's feet.
+
+/obj/machinery/fishtank/proc/destroy(var/deconstruct = FALSE)
+	if(!deconstruct)
+		to_chat(world, "Brutally")														//Check if we are deconstructing or breaking the tank
+		for(var/i in  shard_count)														//Produce the appropriate number of glass shards
+			var/obj/item/weapon/shard/S = new /obj/item/weapon/shard
+			S.forceMove(get_turf(src))
```
Comment (ShiftyRail) [inline code/modules/fish/fish_tank.dm]: Note to reviewers : this doesn't work, the shards don't actually appear

File: code/modules/fish/fish_tank.dm [hunk 16]
```diff
@@ -0,0 +1,685 @@
...
+
+
+/obj/machinery/fishtank/proc/spill_water()
+	switch(tank_type)
+		if("bowl")										//Fishbowl: Wets it's own tile
+			var/turf/T = get_turf(src)
+			if(!istype(T, /turf/simulated))
+				return
+			var/turf/simulated/S = T
+			S.wet = TRUE
```
Comment (ShiftyRail) [inline code/modules/fish/fish_tank.dm]: This may or may not break after @MadmanMartian PR on wetness.
This is also not the correct way to do it at all so we'll be redoing this

**Review by ShiftyRail — COMMENTED** [review]
For information

File: code/modules/fish/fish_tank.dm (hunk 16 above)
Comment (MadmanMartian) [reply code/modules/fish/fish_tank.dm]: Call S.wet(time in deciseconds, TURF_WET_WATER)

File: code/modules/fish/fish_tank.dm (hunk 14 above)
Comment (ShiftyRail) [reply code/modules/fish/fish_tank.dm]: We do, Paradise do not have silicate sprayers.

File: code/modules/fish/fish_tank.dm (hunk 15 above)
Comment (MadmanMartian) [reply code/modules/fish/fish_tank.dm]: Just do new /obj/item/weapon/shard(get_turf(src))

Comment (ShiftyRail) [reply code/modules/fish/fish_tank.dm]: This was the way it was before and it didn't work either.
The culprit is the for loop.

File: code/modules/fish/fish_tank.dm [hunk 17]
```diff
@@ -0,0 +1,685 @@
...
+		if(!dead_fish)								//No fish_item found, possibly due to typo or not being listed. Do nothing.
+			return
+		kill_fish(caught_fish)						//Kill the caught fish from the tank
+		user.visible_message("[user.name] harvests \a [caught_fish] from \the [src].", "You scoop \a [caught_fish] out of \the [src].")
+		new dead_fish(get_turf(user))				//Spawn the appropriate fish_item at the user's feet.
+
+/obj/machinery/fishtank/proc/destroy(var/deconstruct = FALSE)
+	if(!deconstruct)
+		to_chat(world, "Brutally")														//Check if we are deconstructing or breaking the tank
+		for(var/i in  shard_count)														//Produce the appropriate number of glass shards
```
Comment (MadmanMartian) [inline code/modules/fish/fish_tank.dm]: try for(var/i to shard_count) if shard count is going to be an integer

File: code/modules/fish/fish_tank.dm [hunk 18]
```diff
@@ -0,0 +1,685 @@
...
+		dead_fish = fish_items_list[caught_fish]	//Locate the appropriate fish_item for the caught fish
+		if(!dead_fish)								//No fish_item found, possibly due to typo or not being listed. Do nothing.
+			return
+		kill_fish(caught_fish)						//Kill the caught fish from the tank
+		user.visible_message("[user.name] harvests \a [caught_fish] from \the [src].", "You scoop \a [caught_fish] out of \the [src].")
+		new dead_fish(get_turf(user))				//Spawn the appropriate fish_item at the user's feet.
+
+/obj/machinery/fishtank/proc/destroy(var/deconstruct = FALSE)
+	if(!deconstruct)
+		to_chat(world, "Brutally")														//Check if we are deconstructing or breaking the tank
```
Comment (MadmanMartian) [inline code/modules/fish/fish_tank.dm]: Also DEBUG MESSAGE PLS REMOVE

File: code/modules/fish/fish_tank.dm (hunk 13 above)
Comment (ShiftyRail) [reply code/modules/fish/fish_tank.dm]: It doesn't.

File: code/modules/fish/fish_tank.dm (hunk 17 above)
Comment (ShiftyRail) [reply code/modules/fish/fish_tank.dm]: Doesn't work either

File: code/modules/fish/fish_items.dm (hunk 6 above)
Comment (ShiftyRail) [reply code/modules/fish/fish_items.dm]: It is always dropped by the user holding it, since it creates a new obj.
Do you want me to put it in the user's hands ?

File: code/modules/fish/fish_tank.dm [hunk 19]
```diff
@@ -0,0 +1,682 @@
...
+//		ICON PROCS			//
+//////////////////////////////
+
+/obj/machinery/fishtank/update_icon()
+	overlays.Cut()
+
+	//Update Alert Lights
+	if(has_lid)											//Skip the alert lights for aquariums that don't have lids (fishbowls)
+		if(egg_list.len > 0)								//There is at least 1 egg to harvest
+			overlays += "over_egg"
```
Comment (Kurfursten) [inline code/modules/fish/fish_tank.dm]: It seems crazy to me to use overlays to show everything EXCEPT fish

It's a fish tank, you wanna see the fish inside, right?

Comment (ShiftyRail) [reply code/modules/fish/fish_tank.dm]: Most of these overlays are just for the lights
There is currently no sprite of the fish inside the tanks, and I'm afraid I'm not able to do it properly :(

File: code/modules/fish/fish_tank.dm (hunk 13 above)
Comment (Exxion) [reply code/modules/fish/fish_tank.dm]: >`else if`

That said, it should still be replaced with a `switch()`
`switch()` conditions have to be constant expressions, but you can just pass `water_level / water_capacity` instead of multiplying the proportions

Comment (ShiftyRail) [reply code/modules/fish/fish_tank.dm]: @Exxion Can you clarify ?
```
switch(floor(10*water_level/water_capacity))
    if (0.1)
       ...
```
Assuming BYOND has a floor command

Comment (Exxion) [reply code/modules/fish/fish_tank.dm]:
```
switch(water_level / water_capacity)
    if(0.1 to 0.25)
```

The DM Ref is pretty shitty, but you should still at least check it

Comment (ShiftyRail) [reply code/modules/fish/fish_tank.dm]: I r dumb
Thanks

File: code/modules/fish/fish_tank.dm [hunk 20]
```diff
@@ -402,36 +402,38 @@
...
-	else if(water_level <= water_capacity * 0.75)
-		examine_message += "\The [src] is about three-quarters filled. "
-	else if(water_level < water_capacity)
-		examine_message += "\The [src] is nearly full! "
-	else if(water_level == water_capacity)
-		examine_message += "\The [src] is full! "
+	switch (water_level/water_capacity)
+		if (0)
+			examine_message += "\The [src] is empty! "
+		if (0.001 to 0.1)
```
Comment (ShiftyRail) [inline code/modules/fish/fish_tank.dm]: `0.001`
I know what you're thinking, but... it works, up to 0.25u water, which is, I think, the smallest fraction you can get.

**Comment by ShiftyRail** [comment]
This is still alive and in fact, almost finished
I just need to fix an annoying overlay issue and werks the gif @Blithering graciously made for this.

**Comment by ShiftyRail** [comment]
https://streamable.com/y9try
Here is the results with the beautiful fish sprites courtesy of @Blithering, I might need to work them a bit more

**Review by MadmanMartian — APPROVED** [review]

**Comment by Kurfursten** [comment]
>He did the fish overlays

THE ABSOLUTE MADMAN 💯

**Comment by ShiftyRail** [comment]
@Kurfursten Blithering did, I merely coded the overlays and stuff.

**Comment by ShiftyRail** [comment]
https://streamable.com/i2qiw
Shark sprites, courtesy of Blithering again
Might make a few adjustments, also have a bug to fix with examine not showing up properly

Other than that I think there is no shitcode anymore, sprites are there (thought we might have more !) and it works fine. We did it !

File: code/modules/fish/fish_eggs.dm [hunk 21]
```diff
@@ -0,0 +1,83 @@
+/obj/item/fish_eggs
+	name = "fish eggs"
+	desc = "Eggs laid by a fish. This cluster seems... empty?"
+	icon = 'icons/obj/fish_items.dmi'
+	icon_state = "eggs"
+	w_class = 2
```
Comment (DamianX) [inline code/modules/fish/fish_eggs.dm]: Should use defines

File: code/modules/fish/fish_items.dm [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+	icon = 'icons/obj/fish_items.dmi'
+	icon_state = "net"
+	slot_flags = SLOT_BELT
+	throwforce = 0
+	w_class = 2
+	throw_speed = 3
+	throw_range = 7
+
+/obj/item/weapon/fishtools/fish_net/suicide_act(mob/user)			//"A tiny net is a death sentence: it's a net and it's tiny!" https://www.youtube.com/watch?v=FCI9Y4VGCVw
+	to_chat(viewers(user), "<span class='warning'>[user] places the [src] on top of \his head, \his fingers tangled in the netting! It looks like \he's trying to commit suicide.</span>")
```
Comment (DamianX) [inline code/modules/fish/fish_items.dm]: `visible_message`

Comment (DamianX) [inline code/modules/fish/fish_items.dm]: `\the [src]`

File: code/modules/fish/fish_items.dm [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+	desc = "A single raw shrimp."
+	icon = 'icons/obj/seafood.dmi'
+	icon_state = "shrimp_raw"
+	filling_color = "#FF1C1C"
+
+/obj/item/weapon/reagent_containers/food/snacks/shrimp/New()
+	..()
+	desc = pick("Anyway, like I was sayin', shrimp is the fruit of the sea.", "You can barbecue it, boil it, broil it, bake it, saute it.")
+	reagents.add_reagent("NUTRIMENT", 1)
+	src.bitesize = 1
```
Comment (DamianX) [inline code/modules/fish/fish_items.dm]: Why is this in `New`?

File: code/modules/fish/fish_tank.dm [hunk 24]
```diff
@@ -0,0 +1,716 @@
...
+	max_health = 100			// This thing is a freaking wall, it can handle abuse.
+	cur_health = 100
+	shard_count = 3
+
+/obj/machinery/fishtank/wall/Cross(atom/movable/mover, turf/target, height = 1.5, air_group = 0) // Prevents airflow. Copied from windows.
+	if(istype(mover) && mover.checkpass(PASSGLASS))
+		return TRUE
+	return FALSE
+
+/obj/machinery/fishtank/wall/Uncross(atom/movable/O as mob|obj, target as turf)
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: Why is this needed? Default `Uncross` already allows passage

File: code/modules/fish/fish_tank.dm [hunk 25]
```diff
@@ -0,0 +1,716 @@
...
+
+//////////////////////////////
+//		VERBS & PROCS		//
+//////////////////////////////
+
+/obj/machinery/fishtank/verb/toggle_lid_verb()
+	set name = "Toggle Tank Lid"
+	set category = "Object"
+	set src in view(1)
+	if (usr.incapacitated() || usr.isDead() || !usr.Adjacent(src))
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: dead people are also incapacitated

File: code/modules/fish/fish_tank.dm [hunk 26]
```diff
@@ -0,0 +1,716 @@
...
+/obj/machinery/fishtank/proc/add_food(var/amount)
+	food_level = min(MAX_FOOD, food_level + amount)
+
+/obj/machinery/fishtank/proc/check_health()
+	//Max value check
+	if(cur_health > max_health)						//Cur_health cannot exceed max_health, set it to max_health if it does
+		cur_health = max_health
+	//Leaking status check
+	if(cur_health <= (max_health * 0.25))			//Major leak at or below 25% health (-10 water/cycle)
+		leaking = 2
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: magic numbers?

File: code/modules/fish/fish_tank.dm [hunk 27]
```diff
@@ -0,0 +1,716 @@
...
+			if(MAJOR_LEAK)
+				examine_message += "<span class = 'warning'>\The [src] is nearly shattered!</span>"
+
+
+	//Finally, report the full examine_message constructed from the above reports
+	to_chat(user, "[jointext(examine_message, "")]")
+	return examine_message
+
+//////////////////////////////
+//		ATACK PROCS			//
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: attack

File: code/modules/fish/fish_tank.dm [hunk 28]
```diff
@@ -0,0 +1,716 @@
...
+//////////////////////////////
+
+/obj/machinery/fishtank/attack_alien(mob/living/user as mob)
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: as mob

File: code/modules/fish/fish_tank.dm [hunk 29]
```diff
@@ -0,0 +1,716 @@
...
+	if(islarva(user))
+		return
+	attack_generic(user, 15)
+
+
+/obj/machinery/fishtank/attack_hand(mob/user as mob)
+
+	if(user.a_intent == I_HURT)
+		playsound(get_turf(src), 'sound/effects/glassknock.ogg', 80, 1)
+		user.visible_message("<span class='danger'>\The [user] bangs against the [src]!</span>", \
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: `\the [src]`

File: code/modules/fish/fish_tank.dm [hunk 30]
```diff
@@ -0,0 +1,716 @@
...
+	if(cur_health <= 0)
+		user.visible_message("<span class='danger'>\The [user] smashes through \the [src]!</span>")
+		destroy()
+	else	//for nicer text
+		user.visible_message("<span class='danger'>\The [user] smashes into \the [src]!</span>")
+		playsound(loc, 'sound/effects/Glasshit.ogg', 100, 1)
+		check_health()
+
+/obj/machinery/fishtank/attackby(var/obj/item/O, var/mob/user as mob)
+	//Welders repair damaged tanks on help intent, damage on all others
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: you mean silicate sprayers

File: code/modules/fish/fish_tank.dm [hunk 31]
```diff
@@ -0,0 +1,716 @@
...
+	if(issilicatesprayer(O))
+		var/obj/item/device/silicate_sprayer/S = O
+		if(user.a_intent == I_HELP)
+			if (S.get_amount() > 2)
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: should probably be `>=`

File: code/modules/fish/fish_tank.dm [hunk 32]
```diff
@@ -0,0 +1,716 @@
...
+					to_chat(user, "<span class='notice'>You repair some of the cracks on \the [src].</span>")
+					cur_health += 20
+					S.remove_silicate(2)
+					check_health()
+				else
+					to_chat(user, "<span class='warning'>There is no damage to fix!</span>")
+			else if (cur_health < max_health)
+				to_chat(user, "<span class='notice'>You require more silicate to fix the damage on \the [src].</span>")
+		else
+			hit(S.force, user, O)
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: silicate sprayer has `force = 0` anyway, I assume this is a left-over from when it used welders

**Comment by DamianX** [comment]
Many of the review points occur more than once, I didn't leave a note for each of them

**Comment by FalseIncarnate** [comment]
Aquarium code on Paradise has a few updates you seem to be missing here, such as a refactor of how fish breeding is handled and such. Notable changes are the food level requirements for breeding were drastically lowered and some fish will refuse to lay fertile eggs unless a partner of their own species is present, as well as some general code cleanup.

May want to look at the following PRs to see if you want to carry over their changes:
- https://github.com/ParadiseSS13/Paradise/pull/6699
- https://github.com/ParadiseSS13/Paradise/pull/7434

**Comment by ShiftyRail** [comment]
@FalseIncarnate Thanks for the fish datums !
We also did a few improvements here (notably the examine procs), added a few sprites, and used defines most of the time.

File: code/modules/fish/fish_tank.dm [hunk 33]
```diff
@@ -0,0 +1,711 @@
...
+	else
+		switch (leaking)									//No water, report the cracks instead
+			if(MINOR_LEAK)
+				examine_message += "<span class = 'warning'>\The [src] is cracked.</span>"
+			if(MAJOR_LEAK)
+				examine_message += "<span class = 'warning'>\The [src] is nearly shattered!</span>"
+
+
+	//Finally, report the full examine_message constructed from the above reports
+	to_chat(user, "[jointext(examine_message, "")]")
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: why are you wrapping this into a `"[]"`?

File: code/modules/fish/fish_tank.dm [hunk 34]
```diff
@@ -0,0 +1,711 @@
...
+	return examine_message
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: why bother?

File: code/modules/fish/fish_tank.dm [hunk 35]
```diff
@@ -0,0 +1,711 @@
...
+/obj/machinery/fishtank/attack_alien(var/mob/living/user)
+	if(islarva(user))
+		return
+	attack_generic(user, 15)
+
+
+/obj/machinery/fishtank/attack_hand(var/mob/user)
+
+	if(user.a_intent == I_HURT)
+		playsound(get_turf(src), 'sound/effects/glassknock.ogg', 80, 1)
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: Why the inconsistency between the two `playsound `calls? Also you don't need `get_turf`, `playsound` does it

File: code/modules/fish/fish_tank.dm [hunk 36]
```diff
@@ -0,0 +1,711 @@
...
+
+/* tank construction */
+
+/obj/structure/displaycase_frame/attackby(var/obj/item/weapon/F, var/mob/user) // FISH BOWL
+	if (iswelder(F))
+		to_chat(user, "<span class='notice'>You use the machine frame as a vice and shape the glass with the welder into a fish bowl.</span>")
+		getFromPool(/obj/item/stack/sheet/metal, get_turf(src), 5)
+		new /obj/machinery/fishtank/bowl(get_turf(src))
+		qdel(src)
+
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: Should return `TRUE`

File: code/modules/fish/fish_tank.dm [hunk 37]
```diff
@@ -0,0 +1,711 @@
...
+		playsound(src.loc, 'sound/effects/glassknock.ogg', 80, 1)
+		user.visible_message("\The [user] taps on \the [src].", \
+							"You tap on \the [src].", \
+							"You hear a knocking sound.")
+
+
+/obj/machinery/fishtank/proc/hit(var/damage, var/mob/user, var/obj/O)
+	user.delayNextAttack(0.3 SECONDS)
+	user.do_attack_animation(src, O) // Ensuring the legacy lives on :^)
+	playsound(get_turf(src), 'sound/effects/glassknock.ogg', 80, 1)
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: no need for `get_turf`

File: code/modules/fish/fish_tank.dm [hunk 38]
```diff
@@ -0,0 +1,711 @@
...
+							message += "You overfilled \the [src] and some water runs down the side, wasted."
+						C.reagents.clear_reagents()
+				user.visible_message("\The [user] pours the contents of \the [C] into \the [src].", "<span class = 'notice'>[message]</span>")
+				return TRUE
+			//Empty containers will scoop out water, filling the container as much as possible from the water_level
+			else
+				if(water_level == 0)
+					to_chat(user, "<span class='notice'>\The [src] is empty!</span>")
+				else
+					C.reagents.add_reagent("water", water_level)
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: use `WATER`

File: code/modules/fish/fish_tank.dm [hunk 39]
```diff
@@ -0,0 +1,711 @@
...
+			if (BREEDING_THRESHOLD to FOOD_OK)			//Food_level is starting to get low, but still above the breeding threshold
+				overlays += "over_food_somefood"
+			if (FOOD_OK to MAX_FOOD)					//Food_level is below breeding threshold, or fully consumed, feed the fish!
+				overlays += "over_food_fullfood"
+		overlays += "over_leak_[leaking]"				//Green if we aren't leaking, light blue and slow blink if minor link, dark blue and rapid flashing for major leak
+
+	if (fish_list.len)
+		if("shark" in fish_list) // Smaller fish hide when a shark's town
+			switch(tank_type)
+				if ("bowl")
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: Could use defines

File: code/modules/fish/fish_tank.dm [hunk 40]
```diff
@@ -0,0 +1,711 @@
...
+			var/list/L = ST.CardinalTurfs()
+			for(var/turf/T in L)
+				if(!istype(T, /turf/simulated))
+					continue
+				var/turf/simulated/S = T
+				S.wet(10 SECONDS, TURF_WET_WATER)
+
+		if ("wall")										//Wall-tank: Wets it's own tile and the surrounding 8 tiles (3x3 square)
+			for(var/turf/T in view(src, 1))
+				if(!istype(T, /turf/simulated))
```
Comment (DamianX) [inline code/modules/fish/fish_tank.dm]: just filter for `var/turf/simulated/T` in the for loop, same for the other loops

**Review by DamianX — COMMENTED** [review]
Some of the `attackby` branches should return `TRUE`

File: code/modules/fish/fish_tank.dm (hunk 34 above)
Comment (Kurfursten) [reply code/modules/fish/fish_tank.dm]: Could be useful if someone wanted to integrate this into a fishtank monitor or something later.

Comment (DamianX) [reply code/modules/fish/fish_tank.dm]: Surely it would be better to split the logic into a different proc in that case
